### ISSUE TITLE
`tf.reduce_mean` returns wrong results for certain 4D tensor sizes, reductions and distributions

### ISSUE THREAD

**Issue by kretes** [tensorflow/tensorflow#56563]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.8.2, 2.9.1

### Custom Code

No

### OS Platform and Distribution

Ubuntu 18.04.5 LTS

### Python version

3.7 , 3.9

### Current Behaviour?

for tensor of shape `(1, 256, 256, 256, 4)` generated with `tf.random.normal(size, mean=100, stddev=1)` `tf_reduce_mean(tensor, reduction_axes=[0, 1, 2, 3])` returns wrong results.

### Standalone code to reproduce the issue

```shell
mean_val = 100
std = 1

dimensions = (256,)
channels = (4,)

for d in dimensions:
    for ch in channels:    
        size = (1, d, d, d, ch)
        reduction_axes = [i for i in range(len(size) - 1)]

        tensor = tf.random.normal(size, mean=mean_val, stddev=std)                
        mean = tf.reduce_mean(tensor, reduction_axes)
        print("shape:", size, "per-channel mean:", mean.numpy())

        for chann_idx in range(ch):
            single_channel = tensor[..., chann_idx]
            mean = tf.reduce_mean(single_channel)
            print("shape:", single_channel.shape, "channel_idx:", chann_idx, "mean:", mean.numpy())
```
This code snippet shows that the results acquired with `reduce_mean` with `reduction_axes` are wrong. Showing that the right mean is calculated when manually iterating over axis slices.

If you replace the search space to e.g.:
```
dimensions = (64,128,256)
channels = (1,4,8)
```
you may notice some other strange behaviours. e.g. for dimension=128 the error happens for channels=4 but not for 1 or 8.

See https://colab.research.google.com/drive/1o4GdZj6ihpQNawj32N0ukYGX4ngviGRG?usp=sharing
```


### Relevant log output

```shell
shape: (1, 256, 256, 256, 4) per-channel mean: [117.75938  117.759384 117.759384 117.759865]
shape: (1, 256, 256, 256) channel_idx: 0 mean: 99.999664
shape: (1, 256, 256, 256) channel_idx: 1 mean: 100.000015
shape: (1, 256, 256, 256) channel_idx: 2 mean: 100.000305
shape: (1, 256, 256, 256) channel_idx: 3 mean: 99.999626
```
```
</details>

**Comment by tilakrayal** [comment]
Hi @kretes,
I was able to execute the provided code without any issues and also the result is also as expected. Could you find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/8c99de113facb8006da66581a3a7969b/untitled401.ipynb). Thank you!

**Comment by kretes** [comment]
Hi @tilakrayal 
Indeed your colab worked fine.
I now noticed that the issue is when running on CPU, so if you remove the hardware accelerator from your notebook - the behaviour starts to be observable

**Comment by tilakrayal** [comment]
@chunduriv,
I was able to reproduce the issue on `tensorflow-cpu` v2.8, v2.9, nightly and on `tensorflow-gpu` [v2.8](https://colab.research.google.com/gist/tilakrayal/8c99de113facb8006da66581a3a7969b/untitled401.ipynb), the code was executing without any error and the result was also as expected. Kindly find the gist [here](https://colab.research.google.com/gist/tilakrayal/c3168fa6dc7f381fa51972cd433b59df/56563-cpu.ipynb).

**Comment by cantonios** [comment]
Ugh, so I dug into this, and it's because in the first case we're doing a naive linear reduction and eventually run out of precision when adding up elements.

In the second case, when you split the channels, we're doing a more accurate tree reduction.

This is an Eigen "bug" - not bug as in it's wrong, but bug as in it could be doing something more accurate.

**Comment by kretes** [comment]
@cantonios thanks for looking at the issue.

I noticed in the meantime that if I change tf random init:
`tensor = tf.random.normal(size, mean=mean_val, stddev=std)`
to numpy random init:
`tensor = np.random.normal(size=size, loc=mean_val, scale=std)`

I get proper results from the `tf_reduce_mean`.
Does this make sense? See both version in this [gist](https://colab.research.google.com/drive/1MOZNOzqNvEPT0CZbub8-84VnSucHKB86?usp=sharing)

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
